### PR TITLE
Feat/runc checkpoint

### PR DESCRIPTION
**Pull request by bsmithai**
This PR includes a full Containerd dependent checkpoint procedure call process. From client to Containerd to Containerd-shim to Runc, all RPC calls are extracted in order to persist metadata information, leases, namespaces and all other Containerd dependent meta information about checkpoints. Allowing users of Cedana to use OUR container checkpointing on a Containerd managed Runc container.

Note this is a DRAFT PR. Items left to do:
- [x] Re implement our own working runc checkpointing
- [x] Insert our own Runc restore -> groundwork is done so this is fairly easy
- [x] Get rid of any CTR CLI dependencies
- [x] Refactor code for readability

### CODE REVIEW

**Comment by bsmithai** [comment]
Needs restore added to api then we are good

**Comment by bsmithai** [comment]
To try this out:

Install containerd
run: sudo ctr image pull quay.io/quay/busybox:latest
run: sudo ctr run quay.io/quay/busybox:latest test
run: ./cedana daemon
run: ./cedana dump container -i test -p test
grab the printed checkpoint name
run ./cedana restore container -i {checkpointName} -p restoretest

File: cmd/cli.go [hunk 1]
```diff
@@ -77,6 +84,65 @@ var dumpCmd = &cobra.Command{
...
+			WorkDirectory:   workPath,
+			LeaveRunning:    true,
+			TcpEstablished:  false,
+		}
+
+		a := api.RuncDumpArgs{
+			PID:      runcPid,
+			WorkPath: workPath,
+			RuncPath: runcPath,
+			CriuOpts: *criuOpts,
```
Comment (nravic) [inline cmd/cli.go:133]: I think we could use the CriuOpts made in prepareDump here, would be a great first step towards unifying all our checkpointing

File: test/integration/checkpoint.bats
Comment (nravic) [inline test/integration/checkpoint.bats:1]: we'd also need runc in these tests

File: cmd/cli.go [hunk 2]
```diff
@@ -77,6 +84,65 @@ var dumpCmd = &cobra.Command{
...
+			Ref:         ref,
+			ContainerId: containerId,
+		}
+
+		var resp api.ContainerDumpResp
+		err = cli.conn.Call("CedanaDaemon.ContainerDump", a, &resp)
+		if err != nil {
+			return err
+		}
+		cli.logger.Info().Msgf("container %s dumped successfully to %s", containerId, dir)
```
Comment (nravic) [inline cmd/cli.go:107]: Is it actually checkpointing it to a directory? where are the image files stored?

**Review by nravic — COMMENTED** [review]
Looks great so far! small comments, will make another pass after playing with this for a bit

File: test/integration/checkpoint.bats [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+@test "check for ctr" {
+    run sudo ctr version
+    [[ "$status" -eq 0 ]]
+}
+
+@test "pull busybox" {
+    sudo ctr image pull quay.io/quay/busybox:latest
+    sudo setsid ctr run quay.io/quay/busybox:latest testcheckpoint > /dev/null 2>&1 &
```
[the remark this answers is not in the document]
Comment (bsmithai) [reply test/integration/checkpoint.bats:17]: For some reason I couldn't run the command without setsid. It kept taking over the terminal. But I can give it a try again

File: cmd/cli.go (hunk 1 above)
Comment (bsmithai) [reply cmd/cli.go:133]: Good point

File: cmd/cli.go (hunk 2 above)
Comment (bsmithai) [reply cmd/cli.go:107]: It's checkpointing to /tmp/ctd-{uuid}. That's if you don't specify where to unpack. I will add an optional flag for checkpoint directory in the future

Comment (bsmithai) [reply cmd/cli.go:107]: `	if image == "" {
		checkpointImageExists = true
		image, err = os.MkdirTemp(os.Getenv("XDG_RUNTIME_DIR"), "ctrd-checkpoint")
		if err != nil {
			return &apiTasks.CheckpointTaskResponse{}, err
		}
		criuOpts.ImagesDirectory = image
		fmt.Printf("Checkpointing to %s\n", image)
		defer os.RemoveAll(image)
	}`

File: test/integration/checkpoint.bats (hunk 3 above)
[the remark this answers is not in the document]
Comment (nravic) [reply test/integration/checkpoint.bats:17]: oh its a bats thing in that case, not a priority

File: cmd/cli.go (hunk 1 above)
Comment (bsmithai) [reply cmd/cli.go:133]: Can I turn this into a ticket? And proceed for now

File: test/integration/checkpoint.bats [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    [[ "$status" -eq 0 ]]
+}
+
+@test "checkpoint & restore" {
+    output=$(sudo ../../cedana client dump -c testcheckpoint)
```
[the remark this answers is not in the document]
Comment (bsmithai) [reply test/integration/checkpoint.bats:22]: On it boss :+1:

File: cmd/debug.go [hunk 5]
```diff
@@ -41,10 +41,28 @@ var cfgCmd = &cobra.Command{
...
+
+		return nil
+	},
+}
+
+var debugRuncRestoreCmd = &cobra.Command{
+	Use: "runc",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		root := "/var/run/runc"
+		bundle := "/home/brandonsmith/bundle"
```
Comment (nravic) [inline cmd/debug.go:60]: looks like an artifact

File: README.md [hunk 6]
```diff
@@ -47,6 +51,13 @@ See the configuration section for more toggles.
 sudo cedana client restore /path/to/zip
 ```
 
+Currently, we also support `runc` and by extension, `containerd` checkpointing, with more container runtime support planned in the future. Checkpointing these is as simple as prepending the `dump/restore` commands with the correct runtime. For example, to checkpoint a `containerd` container: 
+
+```sh 
+sudo cedana dmp containerd -i test -p test
```
Comment (nravic) [inline README.md:57]: dmp -> dump

File: benchmarking/pids/pytorch.pid
Comment (nravic) [inline benchmarking/pids/pytorch.pid:1]: artifact

File: benchmarking/pids/pytorch_vision.pid
Comment (nravic) [inline benchmarking/pids/pytorch_vision.pid:1]: artifact

**Review by nravic — COMMENTED** [review]
couple more comments, but should be good!!

**Review by nravic — APPROVED** [review]
lgtm :sunglasses:

**Pull request merged**